### PR TITLE
#323 enable `emits_changed_signal` on `dbus_interface`

### PR DESCRIPTION
**Pull request by marcelmasqueprodrive**
Implements  #323: enables interface property macros on getters to be annotated with `emits_changed_signal`.

### CODE REVIEW

**Comment by zeenix** [comment]
Nice! I really liked how you did a great job at splitting your work into atomic commits and added details to commit logs. :+1: It would be awesome if you could also add subcrate prefixes (after the emoji to commit shortlog ( think `zm: ` in your case), then they'll be perfect and consistent with rest of the history. :)

File: zbus_macros/src/iface.rs [hunk 1]
```diff
@@ -62,17 +63,21 @@ impl<'a> Property<'a> {
...
             read: false,
             write: false,
+            emits_changed_signal: PropertyEmitsChangedSignal::True,
             ty: None,
             doc_comments: quote!(),
         }
     }
 }
 
+#[derive(PartialEq)]
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:73]: This change went into the wrong commit. This should be part of the commit adding `MethodType` enum.

File: zbus_macros/src/iface.rs [hunk 2]
```diff
@@ -62,17 +63,21 @@ impl<'a> Property<'a> {
...
 enum MethodType {
     Signal,
     Property(PropertyType),
     Other,
 }
+
+#[derive(PartialEq)]
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:80]: Same here. This change belongs in the previous commit.

File: zbus_macros/src/utils.rs [hunk 3]
```diff
@@ -55,7 +55,7 @@ pub fn is_blank(s: &str) -> bool {
 /// Standard annotation `org.freedesktop.DBus.Property.EmitsChangedSignal`.
 ///
 /// See <https://dbus.freedesktop.org/doc/dbus-specification.html#introspection-format>.
-#[derive(Debug, Default)]
+#[derive(Debug, Default, Clone)]
```
Comment (zeenix) [inline zbus_macros/src/utils.rs:58]: I wouldn't have pointed this one out but since you're going to be fixing the commits, I think best to split this into a separate one too that comes before this ("✨ Adds iface support for EmitsChangedSignal flag.") commit.

File: zbus_macros/tests/tests.rs [hunk 4]
```diff
@@ -220,13 +220,17 @@ fn test_interface() {
     <arg name="arg" type="y"/>
     <arg name="other" type="s"/>
   </signal>
-  <property name="MyCustomProperty" type="u" access="readwrite"/>
+  <property name="MyCustomProperty" type="u" access="readwrite">
+    <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="true"/>
```
Comment (zeenix) [inline zbus_macros/tests/tests.rs:224]: I think `true` is the default? If so, we should skip the annotation when it's `true`. I think other implementations also skip it as otherwise, it will mean a lot of noise in many introspection files.

File: zbus_macros/src/iface.rs [hunk 5]
```diff
@@ -397,8 +395,8 @@ pub fn expand(args: AttributeArgs, mut input: ItemImpl) -> syn::Result<TokenStre
                     // * For argument type with lifetimes, we convert from `Value` (so
                     //   `TryFrom<Value<'_>>` is required).
                     //
-                    // * For all other arg types, we convert the passed value to `OwnedValue` first and
-                    //   then pass it as `Value` (so `TryFrom<OwnedValue>` is required).
+                    // * For all other arg types, we convert the passed value to `OwnedValue` first
+                    //   and then pass it as `Value` (so `TryFrom<OwnedValue>` is required).
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:399]: Is this change needed? If so, it doesn't belong in this commit.

File: zbus_macros/src/iface.rs [hunk 6]
```diff
@@ -185,16 +151,13 @@ pub fn expand(args: AttributeArgs, mut input: ItemImpl) -> syn::Result<TokenStre
         } else {
             None
         };
-
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus_macros/src/iface.rs:188]: You ended up adding a new empty line now. :laughing:

File: zbus_macros/src/iface.rs [hunk 7]
```diff
@@ -208,7 +171,6 @@ pub fn expand(args: AttributeArgs, mut input: ItemImpl) -> syn::Result<TokenStre
         } else {
             quote!(c.reply(m, &reply).await)
         };
-
```
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus_macros/src/iface.rs:211]: Same here. Now you're adding an empty line instead of removing.

File: zbus_macros/src/iface.rs [hunk 8]
```diff
@@ -527,37 +572,75 @@ pub fn expand(args: AttributeArgs, mut input: ItemImpl) -> syn::Result<TokenStre
...
+                            )
                         }
-                    );
+                        PropertyEmitsChangedSignal::Invalidates => {
+                            quote!(
+                                #zbus::fdo::Properties::properties_changed(
+                                    signal_context,
+                                    #zbus::names::InterfaceName::from_static_str_unchecked(#iface_name),
+                                    &::std::collections::HashMap::new(),
+                                    &[],
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:593]: Didn't catch this the last time. The last argument here is supposed to be the list of all invalidated properties. So it think this should be `&[#member_name)],`. It would be great to have a test case for this. Perhaps add the case to both interface and proxy side in `zbus/tests/e2e.rs`?

**Review by zeenix — CHANGES_REQUESTED** [review]
Looking really good otherwise. The request for a testcase is just that: a request. It's ok if you don't have the time to do that, just let me know.

File: zbus_macros/src/iface.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus_macros/src/iface.rs:188]: [looks like the CI wouldn't have let you get away with this anyway](https://github.com/dbus2/zbus/actions/runs/7099835568/job/19327305216?pr=514). :laughing:

File: zbus_macros/tests/tests.rs (hunk 4 above)
Comment (zeenix) [reply zbus_macros/tests/tests.rs]: Did you remove the test all together? :thinking:

File: zbus_macros/src/iface.rs (hunk 8 above)
Comment (marcelmasqueprodrive) [reply zbus_macros/src/iface.rs:591]: This code goes in the `properties_changed_method`, so I do not think that we should be reporting that the property is getting invalidated when we call the `changed` method. However, the `prop_invalidate_method` code should have `&[#member_name]`, this is a few lines below this comment

Comment (zeenix) [reply zbus_macros/src/iface.rs:591]: > This code goes in the `properties_changed_method`, so I do not think that we should be reporting that the property is getting invalidated when we call the `changed` method. However

Not sure I follow. This is as per the behavior defined by [the spec](url):

```
If one or more properties change on an object, the org.freedesktop.DBus.Properties.PropertiesChanged signal may be emitted (this signal was added in 0.14):

              org.freedesktop.DBus.Properties.PropertiesChanged (STRING interface_name,
                                                                 ARRAY of DICT_ENTRY<STRING,VARIANT> changed_properties,
                                                                 ARRAY<STRING> invalidated_properties);
        

where changed_properties is a dictionary containing the changed properties with the new values and invalidated_properties is an array of properties that changed but the value is not conveyed. 
```

Otherwise, the client gets a signal w/o any information at all and hence not useful.

> However, the `prop_invalidate_method` code should have `&[#member_name]`, this is a few lines below this comment

Good that you mentioned. I also missed that you are adding a new separate method for invalidating. I think this is not needed. The method should still be `_changed`. The behavior is then defined by the type of property declared by the user.

File: zbus/tests/e2e.rs [hunk 9]
```diff
@@ -430,6 +463,105 @@ impl MyIfaceImpl {
...
+            modifiable_true: 0,
+            modifiable_invalidates: 0,
+            modifiable_const: 0,
+            modifiable_false: 0,
+        }
+    }
+}
+
+#[dbus_interface(interface = "org.freedesktop.MyEmitsChangedSignalIface")]
+impl MyEmitsChangedSignalIfaceImpl {
```
Comment (zeenix) [inline zbus/tests/e2e.rs:488]: Thanks for adding this. However, I don't think we need to add another interface just for this. Why not add new properties to the existing interface and proxy? You'll basically just need 4-5 new props.

**Review by zeenix — CHANGES_REQUESTED** [review]
Thanks. Sorry to ask you for another round but I think it'll be done soon.

File: zbus/tests/e2e.rs [hunk 10]
```diff
@@ -123,6 +123,36 @@ trait MyIface {
...
+    #[dbus_proxy(property)]
+    fn set_modifiable_default(&self, count: u32) -> zbus::Result<()>;
+
+    #[dbus_proxy(property)]
+    fn modifiable_true(&self) -> zbus::Result<u32>;
+
+    #[dbus_proxy(property)]
+    fn set_modifiable_true(&self, count: u32) -> zbus::Result<()>;
+
+    #[dbus_proxy(property)]
```
Comment (zeenix) [inline zbus/tests/e2e.rs:139]: why isn't the proxy side making use of the (existing) `emits_changed_signal` attributes? Same for other methods.

File: zbus/tests/e2e.rs [hunk 11]
```diff
@@ -699,6 +817,102 @@ async fn my_iface_test(conn: Connection, event: Event) -> zbus::Result<u32> {
...
+            let expected_property_key = "ModifiableInvalidates";
+            let args = changed.args()?;
+            assert_eq!(
+                args.invalidated_properties(),
+                &vec![expected_property_key.to_string()]
+            );
+
+            assert_eq!(args.changed_properties().keys().len(), 0);
+        }
+        None => panic!(""),
```
Comment (zeenix) [inline zbus/tests/e2e.rs:882]: Might as well, just `unwrap` then and reduce code complexity? Same for other matches you're adding here.

**Review by zeenix — CHANGES_REQUESTED** [review]
So sorry, still 2 small things.

**Review by zeenix — APPROVED** [review]
Great job!

**Comment by zeenix** [comment]
> Implements #323

btw, for future reference please use the keywords that GH can pick up on and automatically close the issues (e.g here it a `Fixes #323.` line at the end, would have done the trick.

**Comment by zeenix** [comment]
> Implements #323: enables interface property macros on getters to be annotated with `emits_changed_signal`.

@mmasque For future reference, using the GH keywords (e.g `Fixes #323` here) closes the issue automatically for us.
